mrmustard/lab_dev/sampler.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Samplers for measurement devices.
+"""
+
+from __future__ import annotations
+from itertools import product
+
+from abc import ABC, abstractmethod
+
+from typing import Any, Sequence
+
+import numpy as np
+
+from mrmustard import math, settings
+
+from .states import State, Number, QuadratureEigenstate
+from .circuit_components import CircuitComponent
+from .circuit_components_utils import BtoQ
+
+__all__ = ["Sampler", "PNRSampler", "HomodyneSampler"]
+
+
+class Sampler(ABC):
+    r"""
+    A sampler for measurements of quantum circuits.
+
+    Args:
+        meas_outcomes: The measurement outcomes for this sampler.
+        povms: The POVMs of this sampler.
+    """
+
+    def __init__(
+        self,
+        meas_outcomes: Sequence[Any],
+        povms: CircuitComponent | Sequence[CircuitComponent],
+    ):
+        self._povms = povms
+        self._meas_outcomes = meas_outcomes
+
+    @property
+    def povms(self) -> CircuitComponent | Sequence[CircuitComponent]:
+        r"""
+        The POVMs of this sampler.
+        """
+        return self._povms
+
+    @property
+    def meas_outcomes(self) -> Sequence[Any]:
+        r"""
+        The measurement outcomes of this sampler.
+        """
+        return self._meas_outcomes
+
+    @abstractmethod
+    def probabilities(self, state: State, atol: float = 1e-4) -> Sequence[float]:
+        r"""
+        Returns the probability distribution of a state w.r.t. measurement outcomes.
+
+        Args:
+            state: The state to generate the probability distribution of.
+            atol: The absolute tolerance used for validating the computed probability
+                distribution.
```

Comment:
validating = probabilities sum to 1? Perhaps specify probabilities need to sum to 1 p/m atol. 

This is also related to my question about the state needing to be normalized.